### PR TITLE
fix: backtest

### PR DESCRIPTION
**Pull request by zenixls2**
- fix taker price (not always use m.LastPrice on instant limit trade)
- reorder LastKLine assignment
- reorder interval emit order
- reorder matching engine processing to the beginning of all the kline close event

### CODE REVIEW

File: pkg/backtest/matching.go [hunk 1]
```diff
@@ -180,10 +180,20 @@ func (m *SimplePriceMatching) PlaceOrder(o types.SubmitOrder) (*types.Order, *ty
 	order := m.newOrder(o, orderID)
 
 	if isTaker {
+		var price fixedpoint.Value
 		if order.Type == types.OrderTypeMarket {
 			order.Price = m.Market.TruncatePrice(m.LastPrice)
+			price = order.Price
 		} else if order.Type == types.OrderTypeLimit {
-			order.AveragePrice = m.Market.TruncatePrice(m.LastPrice)
+			if m.LastKLine.High.Compare(order.Price) > 0 && order.Side == types.SideTypeBuy {
```
Comment (c9s) [inline pkg/backtest/matching.go:188]: i think you need to consider the direction of the kline

File: pkg/backtest/matching.go [hunk 2]
```diff
@@ -47,6 +47,12 @@ func init() {
 	}
 }
 
+type Param struct {
+	callback []func(types.KLine, *ExchangeDataSource)
```
Comment (c9s) [inline pkg/backtest/matching.go:51]: use `callbacks` ?

File: pkg/backtest/exchange.go [hunk 3]
```diff
@@ -293,7 +293,7 @@ func (e *Exchange) QueryWithdrawHistory(ctx context.Context, asset string, since
 	return nil, nil
 }
 
-func (e *Exchange) matchingBook(symbol string) (*SimplePriceMatching, bool) {
+func (e *Exchange) MatchingBook(symbol string) (*SimplePriceMatching, bool) {
```
Comment (c9s) [inline pkg/backtest/exchange.go:296]: you don't need to expose this method to public?

File: pkg/backtest/matching.go (hunk 2 above)
Comment (zenixls2) [reply pkg/backtest/matching.go]: done

File: pkg/backtest/exchange.go (hunk 3 above)
Comment (zenixls2) [reply pkg/backtest/exchange.go]: done

File: pkg/backtest/exchange.go [hunk 4]
```diff
@@ -362,21 +362,36 @@ func (e *Exchange) SubscribeMarketData(startTime, endTime time.Time, extraInterv
...
 		}
-
-		// here we generate trades and order updates
-		matching.processKLine(k)
+		matching.ParamCache = make(map[types.Interval]Param)
+	}
+	matching.ParamCache[k.Interval] = Param{
+		callbacks: handlers,
+		src:       src,
+		kline:     k,
```
Comment (c9s) [inline pkg/backtest/exchange.go:393]: there will be a delay for intervals like 1h, 2h, 4h

1h kline closed event will be delayed for 1h
2h will be delayed for 2h
4h will be delayed for 4h

File: pkg/backtest/exchange.go [hunk 5]
```diff
@@ -362,21 +362,36 @@ func (e *Exchange) SubscribeMarketData(startTime, endTime time.Time, extraInterv
...
 
-func (e *Exchange) ConsumeKLine(k types.KLine) {
-	if k.Interval == types.Interval1m {
-		e.currentTime = k.EndTime.Time()
-
-		matching, ok := e.matchingBook(k.Symbol)
-		if !ok {
-			log.Errorf("matching book of %s is not initialized", k.Symbol)
-			return
+func (e *Exchange) ConsumeKLine(k types.KLine, handlers []func(types.KLine, *ExchangeDataSource), src *ExchangeDataSource) {
```
Comment (c9s) [inline pkg/backtest/exchange.go:365]: Is there any way to make this cleaner? I don't want to move the kline handlers into this method, it makes it harder to be refactored.

Comment (zenixls2) [reply pkg/backtest/exchange.go:365]: you could:
1. either expand the caching code in cmd/backtest.go
2. add api to query handlers from ExchangeDataSource during the runtime

File: pkg/backtest/matching.go [hunk 6]
```diff
@@ -47,6 +47,12 @@ func init() {
 	}
 }
 
+type Param struct {
+	callbacks []func(types.KLine, *ExchangeDataSource)
```
Comment (c9s) [inline pkg/backtest/matching.go:51]: move this callback to ExchangeDataSource?

File: pkg/backtest/exchange.go [hunk 7]
```diff
@@ -362,21 +362,36 @@ func (e *Exchange) SubscribeMarketData(startTime, endTime time.Time, extraInterv
...
+	matching, ok := e.matchingBook(k.Symbol)
+	if !ok {
+		log.Errorf("matching book of %s is not initialized", k.Symbol)
+		return
+	}
+	if matching.ParamCache == nil {
```
Comment (c9s) [inline pkg/backtest/exchange.go:371]: make this field ParamCache private?

**Review by narumiruna — APPROVED** [review]

File: pkg/backtest/matching.go [hunk 8]
```diff
@@ -180,10 +188,20 @@ func (m *SimplePriceMatching) PlaceOrder(o types.SubmitOrder) (*types.Order, *ty
 	order := m.newOrder(o, orderID)
 
 	if isTaker {
+		var price fixedpoint.Value
 		if order.Type == types.OrderTypeMarket {
 			order.Price = m.Market.TruncatePrice(m.LastPrice)
+			price = order.Price
 		} else if order.Type == types.OrderTypeLimit {
-			order.AveragePrice = m.Market.TruncatePrice(m.LastPrice)
+			if m.NextKLine != nil && m.NextKLine.High.Compare(order.Price) > 0 && order.Side == types.SideTypeBuy {
```
Comment (c9s) [inline pkg/backtest/matching.go:196]: Can you add a comment describing this behavior?

File: pkg/backtest/exchange.go (hunk 7 above)
Comment (zenixls2) [reply pkg/backtest/exchange.go]: done

File: pkg/backtest/matching.go (hunk 6 above)
Comment (zenixls2) [reply pkg/backtest/matching.go]: done

File: pkg/backtest/matching.go (hunk 8 above)
Comment (zenixls2) [reply pkg/backtest/matching.go:194]: done

File: pkg/backtest/exchange.go [hunk 9]
```diff
@@ -363,20 +365,35 @@ func (e *Exchange) SubscribeMarketData(startTime, endTime time.Time, extraInterv
...
+			panic("expect 1m kline, get " + kline1m.Interval.String())
 		}
-
+		e.currentTime = kline1m.EndTime.Time()
 		// here we generate trades and order updates
-		matching.processKLine(k)
+		matching.processKLine(kline1m)
+		matching.NextKLine = &k
+		for _, kline := range matching.klineCache {
+			// log.Errorf("kline %v, next %v", param.kline, matching.NextKLine)
```
Comment (c9s) [inline pkg/backtest/exchange.go:387]: remove debug log?

File: pkg/fixedpoint/dec.go [hunk 10]
```diff
@@ -1061,6 +1061,9 @@ func (v *Value) UnmarshalYAML(unmarshal func(a interface{}) error) (err error) {
 
 // FIXME: should we limit to 8 prec?
 func (v Value) MarshalJSON() ([]byte, error) {
+	if v.IsInf() {
+		return []byte("\"" + v.String() + "\""), nil
```
Comment (c9s) [inline pkg/fixedpoint/dec.go:1065]: you can use strconv.Quote here

File: pkg/cmd/backtest.go [hunk 11]
```diff
@@ -438,6 +438,9 @@ var BacktestCmd = &cobra.Command{
 		}
 
 		runCtx, cancelRun := context.WithCancel(ctx)
+		for _, exK := range exchangeSources {
```
Comment (c9s) [inline pkg/cmd/backtest.go:441]: when we have 2 exchange sources, these callbacks will be triggered twice?

File: pkg/backtest/exchange.go [hunk 12]
```diff
@@ -363,20 +365,35 @@ func (e *Exchange) SubscribeMarketData(startTime, endTime time.Time, extraInterv
...
+	}
+	if matching.klineCache == nil {
+		matching.klineCache = make(map[types.Interval]types.KLine)
+	}
 
-		matching, ok := e.matchingBook(k.Symbol)
-		if !ok {
-			log.Errorf("matching book of %s is not initialized", k.Symbol)
-			return
+	kline1m, ok := matching.klineCache[k.Interval]
```
Comment (c9s) [inline pkg/backtest/exchange.go:377]: how to make sure the k.Interval is 1m ?

File: pkg/fixedpoint/dec.go (hunk 10 above)
Comment (zenixls2) [reply pkg/fixedpoint/dec.go:1065]: We don't need to escape the quotation, seems like overkilled

File: pkg/backtest/exchange.go (hunk 12 above)
Comment (zenixls2) [reply pkg/backtest/exchange.go:377]: In kline query sql, we ensure that kline with same endtime will be sorted by start time in decreasing order. Therefore for every time frame, 1m kline will be called in priority and trigger klineCache pop.
We could ensure this behaviour unless there's a missing kline1m (and should cause panic to notify developers to backfill the missing data)

File: pkg/cmd/backtest.go (hunk 11 above)
Comment (zenixls2) [reply pkg/cmd/backtest.go:441]: Yes, one for each source

File: pkg/backtest/exchange.go (hunk 9 above)
Comment (zenixls2) [reply pkg/backtest/exchange.go]: done

File: pkg/backtest/exchange.go [hunk 13]
```diff
@@ -363,20 +365,35 @@ func (e *Exchange) SubscribeMarketData(startTime, endTime time.Time, extraInterv
...
+	if ok { // pop out all the old
+		if kline1m.Interval != types.Interval1m {
+			panic("expect 1m kline, get " + kline1m.Interval.String())
```
[the remark this answers is not in the document]
Comment (zenixls2) [reply pkg/backtest/exchange.go]: done

**Review by c9s — APPROVED** [review]

**Pull request merged**